src/Raven.Server/Documents/Queries/Results/TimeSeriesRetriever.cs

```diff
@@ -870,5 +890,107 @@ private static DynamicJsonValue AddTimeSeriesResult(TimeSeriesFunction func, Tim
                 return dt;
             }
         }
+
+        private class MultiReader
+        {
+            private readonly TimeSeriesRetriever _retriever;
+            private TimeSeriesStorage.Reader _reader;
+            private readonly string _docId, _source;
+            private readonly TimeSpan? _offset;
+            private SortedList<long, string> _names;
+            private int _current;
+
+            public MultiReader(TimeSeriesRetriever retriever, string documentId,
+                string source, TimeSpan? offset)
+            {
+                if (string.IsNullOrEmpty(source)) 
+                    throw new ArgumentNullException(nameof(source));
+                _source = source;
+                _docId = documentId ?? throw new ArgumentNullException(nameof(documentId));
+                _offset = offset;
+                _current = 0;
+                _retriever = retriever;
```

Comment:
Can we decouple this from the `retriever`? and pass `collection` and `context` instead?

Reply:
done